stackslib/src/chainstate/stacks/boot/pox-4-vote.clar

```diff
@@ -0,0 +1,77 @@
+;;
+;; @contract voting for the aggregate public key
+;;
+
+;; maps dkg round and signer to proposed aggregate public key
+(define-map votes {reward-cycle: uint, round: uint, signer: principal} {aggregate-public-key: (buff 33), reward-slots: uint})
+;; maps dkg round and aggregate public key to weights of signers supporting this key so far
+(define-map tally {reward-cycle: uint, round: uint, aggregate-public-key: (buff 33)} uint)
+;; maps aggregate public keys to rewards cycles and rounds
+(define-map used-aggregate-public-keys (buff 33) {reward-cycle: uint, round: uint})
+
+(define-constant err-not-allowed (err u10000))
+(define-constant err-incorrect-reward-cycle (err u10001))
+(define-constant err-old-round (err u10002))
+(define-constant err-invalid-aggregate-public-key (err u10003))
+(define-constant err-duplicate-aggregate-public-key (err u10004))
+(define-constant err-duplicate-vote (err u10005))
+(define-constant err-invalid-burn-block-height (err u10006))
+
+(define-constant pox-info
+    (unwrap-panic (contract-call? .pox-4 get-pox-info)))
+
+;; maps reward-cycle ids to last round
+(define-map rounds uint uint)
+
+(define-data-var state-1 {reward-cycle: uint, round: uint, aggregate-public-key: (optional (buff 33)),
+    total-votes: uint}  {reward-cycle: u0, round: u0, aggregate-public-key: none, total-votes: u0})
+(define-data-var state-2 {reward-cycle: uint, round: uint, aggregate-public-key: (optional (buff 33)),
+    total-votes: uint}  {reward-cycle: u0, round: u0, aggregate-public-key: none, total-votes: u0})
+
+;; get voting info by burn block height
+(define-read-only (get-info (height uint))
+    (ok (at-block (unwrap! (get-block-info? id-header-hash height) err-invalid-burn-block-height) (get-current-info))))
+
+;; get current voting info
+(define-read-only (get-current-info)
+    (if (is-eq (/ (current-reward-cycle) u2) u0) (var-get state-1) (var-get state-2)))
+
+(define-read-only (burn-height-to-reward-cycle (height uint))
+    (/ (- height (get first-burnchain-block-height pox-info)) (get reward-cycle-length pox-info)))
+
+(define-read-only (current-reward-cycle)
+    (burn-height-to-reward-cycle burn-block-height))
+    
+(define-read-only (get-last-round (reward-cycle uint))
+    (map-get? rounds reward-cycle))
+
+(define-read-only (get-vote (reward-cycle uint) (round uint) (signer principal))
+    (map-get? votes {reward-cycle: reward-cycle, round: round, signer: signer}))
+    
+(define-read-only (get-signer-slots (signer principal) (reward-cycle uint))
+    (contract-call? .signers get-signer-slots signer reward-cycle))
+
+;; aggregate public key must be unique and can be used only in a single cycle-round pair
+(define-read-only (is-valid-aggregated-public-key (key (buff 33)) (dkg-id {reward-cycle: uint, round: uint}))
+    (is-eq (default-to dkg-id (map-get? used-aggregate-public-keys key)) dkg-id))
+
+(define-public (vote-for-aggregate-public-key (key (buff 33)) (reward-cycle uint) (round uint))
```

Comment:
See comment about `stackerdb-get-signer-slots` in `.signers`. If we use that pattern then instead of a `(buff 33)` key you can accept a signer index `uint` here.

Reply:
I don't understand. Is the `key` related to the signer index?

Reply:
The signers know their own index so when they cast a vote by calling into the contract they can provide that index. You can then retrieve the tuple from `.signers` using that index and do your checks.

Reply:
using index now